### PR TITLE
feat: expose_secrets param on /users/me + sandbox-scoped secrets API

### LINKED ISSUE #13268: feat: SDK-created conversations should inherit SaaS settings (credentials, repo context)
## Problem

When using the SDK with OpenHands Cloud, there are two ways to create a conversation — and **neither gives you both SaaS credentials AND agent customization**:

| | SaaS credentials (GitHub App token, LLM key) | Custom agent (tools, prompt, skills) | Visible in Cloud UI |
|---|---|---|---|
| **Path 1**: `POST /api/v1/app-conversations` | ✅ | ❌ (hardcoded default agent) | ✅ |
| **Path 2**: `OpenHandsCloudWorkspace` + SDK `Conversation` | ❌ (must supply your own) | ✅ | ❌ (see #13272) |

We need a way for SDK-created conversations to inherit the user's SaaS settings (credentials, repo context) while retaining full control over the agent.

### Script A: SaaS credentials ✅, custom agent ❌

Using the `app-conversations` API gives you Cloud credentials and visibility, but you're stuck with the platform's default agent — no way to pass custom tools, system prompt, or skills.

```python
"""Path 1: app-conversations API — gets SaaS credentials, loses agent customization.

Requires: OPENHANDS_API_KEY
"""
import os
import httpx

CLOUD_URL = "https://app.all-hands.dev"
API_KEY = os.environ["OPENHANDS_API_KEY"]

with httpx.Client(timeout=60) as client:
    resp = client.post(
        f"{CLOUD_URL}/api/v1/app-conversations",
        headers={"Authorization": f"Bearer {API_KEY}"},
        json={
            "selected_repository": "OpenHands/OpenHands",
            "git_provider": "github",
            "initial_message": {"message": "Say hello"},
            # ❌ No way to specify:
            #   - custom tools
            #   - custom system prompt
            #   - custom skills / agent context
            #   - CLI mode vs browser mode
            # The platform builds a hardcoded default agent internally.
        },
    )
    resp.raise_for_status()
    task = resp.json()
    print(f"Conversation: {task.get('conversation_id')}")
    print("✅ Uses Cloud's GitHub App token, LLM key, repo cloned")
    print("✅ Visible in Cloud dashboard")
    print("❌ Cannot customize agent (tools, prompt, skills)")
```

### Script B: Custom agent ✅, SaaS credentials ❌

Following the [official SDK example](https://docs.openhands.dev/sdk/guides/agent-server/cloud-workspace#openhands-cloud-workspace), you get full agent control but lose all SaaS integration.

```python
"""Path 2: OpenHandsCloudWorkspace — full agent control, no SaaS credentials.

Requires: OPENHANDS_API_KEY, LLM_API_KEY

Follows: https://docs.openhands.dev/sdk/guides/agent-server/cloud-workspace
"""
import os
from pydantic import SecretStr
from openhands.sdk import LLM, Conversation
from openhands.tools.preset.default import get_default_agent
from openhands.workspace import OpenHandsCloudWorkspace

# ❌ Must provide your own LLM key — Cloud's key is not available
llm = LLM(
    model="anthropic/claude-sonnet-4-5-20250929",
    api_key=SecretStr(os.environ["LLM_API_KEY"]),
)

# ✅ Full control over agent configuration
agent = get_default_agent(llm=llm, cli_mode=True)

with OpenHandsCloudWorkspace(
    cloud_api_url="https://app.all-hands.dev",
    cloud_api_key=os.environ["OPENHANDS_API_KEY"],
) as workspace:
    # ❌ Sandbox has no repo cloned, no GitHub App token, no setup scripts run
    conversation = Conversation(agent=agent, workspace=workspace)
    conversation.send_message("Say hello")
    conversation.run()
    # ❌ Conversation is invisible in Cloud dashboard (see #13272)
```

## Desired Outcome

A single path where SDK users can:
1. **Inherit SaaS credentials** — the sandbox gets the user's GitHub App token, LLM key, and provider tokens from their Cloud account
2. **Inherit repo context** — a specified repository is cloned with setup scripts executed
3. **Customize the agent** — full SDK control over tools, system prompt, skills, and plugins
4. **Be visible in Cloud UI** — conversations appear in the dashboard with full observability

## Relationship to #13272

This is a **follow-up to #13272** (conversations invisible in Cloud UI). The solution to #13272 will determine how conversations get registered with the Cloud, which directly informs how we solve credential/settings inheritance here. For example:

- If #13272 is solved by **auto-registering** every conversation created on a Cloud-provisioned sandbox, then this issue becomes about injecting SaaS credentials into those sandboxes (either automatically or via an explicit SDK call).
- If #13272 is solved by providing an **explicit registration API**, then we could extend that same API to also pull in the user's SaaS settings at registration time.

Either way, the architecture chosen in #13272 shapes the solution here.

## Current Workaround

SDK cloud mode requires users to provide their own `LLM_API_KEY` and `GITHUB_TOKEN`. Reviews are posted under the token owner's identity (not the Cloud bot), and conversations may not appear in the Cloud dashboard.

### PR DESCRIPTION
**Pull request by xingyaoww**
## Summary

Two changes enabling SDK-created conversations to inherit the user's SaaS credentials:

1. **`expose_secrets` query param on `GET /api/v1/users/me`** — when `true`, `llm_api_key` (and other `SecretStr` fields) are returned unmasked. **Requires a valid `X-Session-API-Key` header** for an active sandbox owned by the authenticated user.

2. **Sandbox-scoped secrets endpoints** — authenticated via `X-Session-API-Key` for agent-servers running inside sandboxes to retrieve user secrets on demand.

## Security

`expose_secrets=true` is gated behind dual authentication:
- **Bearer token** (proves identity via `OPENHANDS_API_KEY`)
- **`X-Session-API-Key` header** (proves the caller has an active sandbox)

The server validates that the session key maps to a sandbox owned by the Bearer token user. A leaked API key alone **cannot** extract raw LLM credentials — an ephemeral, sandbox-scoped session key is also required.

## Endpoints

| Endpoint | Auth | Returns | Called by |
|---|---|---|---|
| `GET /api/v1/users/me?expose_secrets=true` | Bearer + `X-Session-API-Key` | Full user settings with unmasked secrets | SDK client (`workspace.get_llm()`) |
| `GET /api/v1/users/me` | Bearer | Full user settings (secrets masked) | Frontend / general use |
| `GET /api/v1/sandboxes/{id}/settings/secrets` | `X-Session-API-Key` | List of secret names (no values) | SDK client (`workspace.get_secrets()`) |
| `GET /api/v1/sandboxes/{id}/settings/secrets/{name}` | `X-Session-API-Key` | Raw secret value (plain text) | Agent-server **inside** sandbox (via `LookupSecret`) |

## Companion PR

- SDK: OpenHands/software-agent-sdk#2409

## Tests

9 unit tests: 4 for `/users/me` expose_secrets (happy path + missing key + wrong user + default masking) + 5 for secrets endpoints.

Resolves #13268

---

To run this PR locally, use the following command:

GUI with Docker:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.openhands.dev/openhands/runtime:300fe82-nikolaik   --name openhands-app-300fe82   docker.openhands.dev/openhands/openhands:300fe82
```

### CODE REVIEW

File: openhands/app_server/sandbox/sandbox_router.py [hunk 1]
```diff
@@ -94,3 +113,102 @@ async def delete_sandbox(
...
+    ),
+) -> SandboxInfo:
+    """Authenticate via ``X-Session-API-Key`` and verify sandbox ownership."""
+    if not session_api_key:
+        raise HTTPException(
+            status.HTTP_401_UNAUTHORIZED,
+            detail='X-Session-API-Key header is required',
+        )
+
+    state = InjectorState()
```
Comment (xingyaoww) [inline openhands/app_server/sandbox/sandbox_router.py]: @OpenHands is the use of InjectorState here an established pattern in this codebase?

**Comment by xingyaoww** [comment]
@OpenHands can you address these comments: https://github.com/OpenHands/OpenHands/pull/13383#pullrequestreview-3945435683

File: openhands/app_server/sandbox/sandbox_router.py (hunk 1 above)
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py]: Yes, this is an established pattern — see `webhook_router.py:valid_sandbox()` which uses the same `InjectorState() + ADMIN` approach. The sandbox service is user-scoped, so looking up a sandbox by session key (which could belong to any user) requires an admin context. The session key itself acts as the proof of access.

In 86a79b8c9 I extracted this into a shared `session_auth.validate_session_key()` utility with documentation explaining the pattern, so it now lives in one place with clear comments.

[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py]: Addressed in 86a79b8c9. The `InjectorState + ADMIN` pattern is established in `webhook_router.py:valid_sandbox()` — the sandbox service is user-scoped, so cross-user lookup by session key requires admin context. The session key itself is the access proof.

I have extracted this into a shared `session_auth.validate_session_key()` utility with clear documentation explaining the rationale, so it now lives in exactly one place.

File: openhands/app_server/user/user_router.py [hunk 2]
```diff
@@ -18,9 +25,65 @@
...
     """Get the current authenticated user."""
     user = await user_context.get_user_info()
     if user is None:
         raise HTTPException(status.HTTP_401_UNAUTHORIZED, detail='Not authenticated')
+    if expose_secrets:
+        await _validate_session_key_ownership(user_context, x_session_api_key)
+        return JSONResponse(  # type: ignore[return-value]
+            content=user.model_dump(mode='json', context={'expose_secrets': True})
+        )
     return user
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/user/user_router.py:44]: Fixed in 86a79b8c9. Extracted `validate_session_key()` into `openhands/app_server/sandbox/session_auth.py`. Both `sandbox_router.py` and `user_router.py` now import and use this shared utility instead of duplicating the logic.

File: openhands/app_server/user/user_router.py [hunk 3]
```diff
@@ -18,9 +25,65 @@
...
+    # Verify the sandbox is owned by the authenticated user
+    caller_id = await user_context.get_user_id()
+    if sandbox_info.created_by_user_id and caller_id:
+        if sandbox_info.created_by_user_id != caller_id:
+            _logger.warning(
+                'Session key user mismatch: sandbox owner=%s, caller=%s',
+                sandbox_info.created_by_user_id,
+                caller_id,
+            )
+            raise HTTPException(
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/user/user_router.py]: Fixed in 86a79b8c9. The ownership check is now explicit:

```python
caller_id = await user_context.get_user_id()
if not caller_id:
    raise HTTPException(401, "Cannot determine authenticated user")

if sandbox_info.created_by_user_id != caller_id:
    raise HTTPException(403, ...)
```

A `None` caller_id now raises 401 instead of silently succeeding. Added a test (`test_expose_secrets_rejects_unknown_caller`) verifying this behavior.

File: tests/unit/app_server/test_sandbox_secrets_router.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
+"""Unit tests for the sandbox settings endpoints and /users/me expose_secrets.
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply tests/unit/app_server/test_sandbox_secrets_router.py]: Partially addressed in 86a79b8c9. Added 6 new unit tests for the shared `validate_session_key()` utility covering: missing key, empty key, invalid key, valid key, and SAAS mode without user. Also added a test for the unknown caller_id case.

The test suite now has 15 tests total (up from 9). Full integration tests (real sandboxes via the API) are out of scope for this PR since the test infrastructure for spinning up real sandboxes is not set up in the unit test suite, but these unit tests do verify all the auth boundary conditions at the function level.

**Comment by xingyaoww** [comment]
@OpenHands can you help me address all the reviews in this PR?

File: openhands/app_server/sandbox/sandbox_router.py [hunk 5]
```diff
@@ -94,3 +108,72 @@ async def delete_sandbox(
...
+) -> SandboxInfo:
+    """Authenticate via ``X-Session-API-Key`` and verify sandbox ownership."""
+    sandbox_info = await validate_session_key(session_api_key)
+
+    if sandbox_info.id != sandbox_id:
+        raise HTTPException(
+            status.HTTP_403_FORBIDDEN,
+            detail='Session API key does not match sandbox',
+        )
+
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py:133]: Fixed in 6a335c857. Replaced `assert sandbox_info.created_by_user_id` with `if not sandbox_info.created_by_user_id: raise HTTPException(401, "Sandbox has no associated user")`. This ensures the check cannot be bypassed under `python -O`.

[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py:133]: Fixed in 6a335c857. Same fix as above — replaced `assert` with explicit `HTTPException(401)`.

[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py:133]: Fixed in 6a335c857. Replaced `assert sandbox_info.created_by_user_id` with `if not sandbox_info.created_by_user_id: raise HTTPException(401)`. This ensures the check cannot be bypassed under `python -O`.

[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py:133]: Fixed in 6a335c857. Same fix as above.

File: tests/unit/app_server/test_sandbox_secrets_router.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply tests/unit/app_server/test_sandbox_secrets_router.py]: Addressed in 6a335c857. Added 9 integration tests using FastAPI TestClient that make real HTTP requests through the full auth validation stack. Only the data layer (sandbox service, user context) is mocked — validate_session_key, _validate_session_key_ownership, and ownership checks all run unmodified. Tests cover: Bearer only (401), invalid session key (401), wrong user (403), valid dual-auth (200 with unmasked secrets), default masking, sandbox secrets without key (401), invalid key (401), mismatched sandbox ID (403), and sandbox without user (401).

File: tests/unit/app_server/test_sandbox_secrets_router.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+Tests:
+- GET /api/v1/users/me?expose_secrets=true
+- GET /api/v1/sandboxes/{sandbox_id}/settings/secrets
+- GET /api/v1/sandboxes/{sandbox_id}/settings/secrets/{secret_name}
+- Shared session_auth.validate_session_key()
+"""
+
+from unittest.mock import AsyncMock, patch
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply tests/unit/app_server/test_sandbox_secrets_router.py]: Addressed in 6a335c857. Same fix as above — added 9 integration tests via TestClient.

File: AGENTS.md [hunk 8]
```diff
@@ -342,3 +342,22 @@ To add a new LLM model to OpenHands, you need to update multiple files across bo
 - Models appear in CLI provider selection based on the verified arrays
 - The `organize_models_and_providers` function groups models by provider
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply AGENTS.md:343]: Fixed in 6a335c857. Added documentation for `GET /api/v1/users/me?expose_secrets=true` including auth requirements and SDK usage context.

File: AGENTS.md [hunk 9]
```diff
@@ -342,3 +342,22 @@ To add a new LLM model to OpenHands, you need to update multiple files across bo
...
 - Default model selection prioritizes verified models for each provider
+
+### Sandbox Settings API (SDK Credential Inheritance)
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply AGENTS.md:346]: Fixed in 6a335c857. Same fix as above.

File: openhands/app_server/user/user_router.py [hunk 10]
```diff
@@ -18,9 +24,52 @@
...
+    ),
+    x_session_api_key: str | None = Header(default=None),
 ) -> UserInfo:
     """Get the current authenticated user."""
     user = await user_context.get_user_info()
     if user is None:
         raise HTTPException(status.HTTP_401_UNAUTHORIZED, detail='Not authenticated')
+    if expose_secrets:
+        await _validate_session_key_ownership(user_context, x_session_api_key)
+        return JSONResponse(  # type: ignore[return-value]
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/app_server/user/user_router.py:41]: Acknowledged. The JSONResponse pattern is documented in the AGENTS.md section and the integration tests verify both masked and unmasked serialization paths.

**Comment by xingyaoww** [comment]
## ✅ Integration Test Report — SDK credential inheritance (2026-03-16)

**SDK PR:** [software-agent-sdk#2409](https://github.com/OpenHands/software-agent-sdk/pull/2409)
**SaaS server:** Custom build from this PR (provides `/sandboxes/{id}/settings/secrets` endpoints)
**Staging:** `ohpr-13383-240.staging.all-hands.dev`
**Test artifact:** [`.pr/README.md`](https://github.com/OpenHands/software-agent-sdk/blob/feat/cloud-workspace-get-llm-secrets/.pr/README.md)

### Results

| Component | Status | Details |
|---|---|---|
| Sandbox provisioning | ✅ | Created, RUNNING in ~50s, cleaned up on exit |
| `workspace.get_llm()` | ✅ | `litellm_proxy/minimax-m2.5` + api_key + base_url from SaaS |
| `workspace.get_secrets()` | ✅ | Discovers `["DUMMY_1", "DUMMY_2"]` via `GET /sandboxes/{id}/settings/secrets` |
| `update_secrets(LookupSecret)` | ✅ | LookupSecret with session key in `headers` survives serialization |
| Env vars exported inside sandbox | ✅ | `_export_envs` resolves LookupSecret → secrets appear as real env vars |
| Sandbox cleanup | ✅ | `DELETE /api/v1/sandboxes/{id}` succeeds |

### Agent output from inside the sandbox

```
DUMMY_1: len=14, last_half=ecret 1
DUMMY_2: len=14, last_half=ecret 2
```

Both secrets are **14 chars, non-empty**. Second half matches expected values (`"Dummy secret 1"` → `"ecret 1"`, `"dummy secret 2"` → `"ecret 2"`).

### Issues found and fixed (in SDK PR)

1. **Sandbox DELETE 405** — `cleanup()` hit wrong endpoint. Fixed to `DELETE /sandboxes/{id}`.

2. **SecretStr serialization redacts headers** — `model_dump(mode="json")` redacted `X-Session-API-Key` in LookupSecret headers. Fixed by adding `context={"expose_secrets": True}` in `update_secrets()`.

3. **Removed `env_headers`** — Original design stored header→env_var mappings so the session key VALUE never appeared in serialized JSON. This created a deployment dependency (sandbox agent-server needed the new field). Dropped it — using existing `headers` field with `expose_secrets` context works with stock sandbox agent-server. **-35 lines, +10 lines.**

File: openhands/app_server/user/user_router.py [hunk 11]
```diff
@@ -18,9 +24,52 @@
 @router.get('/me')
 async def get_current_user(
     user_context: UserContext = user_dependency,
+    expose_secrets: bool = Query(
+        default=False,
+        description='If true, return unmasked secret values (e.g. llm_api_key). '
+        'Requires a valid X-Session-API-Key header for an active sandbox '
+        'owned by the authenticated user.',
+    ),
```
Comment (malhotra5) [inline openhands/app_server/user/user_router.py:32]: NIT: could we separate this into a sandbox route instead? (e.g `@router.get('/{sandbox_id}/settings')`)

Reason is two fold

1. it separates the sandbox routes more obviously from this one. we'll also only require session api key for sandbox routes ideally, i think dual auth is not required
2. we can return more specific data, or extend this endpoint in the future to return configurations that don't live in user settings but should be part of the cloud conversation (e.g some org level config)

File: openhands/app_server/sandbox/sandbox_router.py [hunk 12]
```diff
@@ -94,3 +108,76 @@ async def delete_sandbox(
...
+async def get_secret_value(
+    secret_name: str,
+    sandbox_info: SandboxInfo = Depends(_valid_sandbox_from_session_key),
+) -> Response:
+    """Return a single secret value as plain text.
+
+    Called by ``LookupSecret`` inside the sandbox.
+    """
+    user_context = await _get_user_context(sandbox_info)
+    secret_sources = await user_context.get_secrets()
```
Comment (malhotra5) [inline openhands/app_server/sandbox/sandbox_router.py]: Thoughts on also looking up across `get_provider_tokens` here to support fetching provider secrets like `GITHUB_TOKEN`? Or maybe we don't want to enable that by default?

**Review by malhotra5 — APPROVED** [review]
LGTM!

File: openhands/app_server/sandbox/sandbox_router.py (hunk 12 above)
Comment (malhotra5) [reply openhands/app_server/sandbox/sandbox_router.py]: we can `user_context.get_provider_tokens()` as here as well!

File: openhands/app_server/user/user_router.py (hunk 11 above)
Comment (xingyaoww) [reply openhands/app_server/user/user_router.py:32]: > _This reply was posted by OpenHands AI on behalf of @xingyaoww, based on our Mar 17 sync meeting notes._

Thanks for the suggestion! We discussed this in our sync — I initially had it as a separate sandbox route but ended up duplicating all the logic. The other consideration is that even for sandbox-specific features (e.g. org-level automation configs), the SaaS UI will also need the same information for visualization, so it makes sense to keep a single path for now.

Agreed we can rework this later if it becomes annoying. Keeping as-is for this PR.

File: openhands/app_server/sandbox/sandbox_router.py (hunk 12 above)
Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py]: > _This reply was posted by OpenHands AI on behalf of @xingyaoww, based on our Mar 17 sync meeting notes._

Good call — this is definitely a gap. GITHUB_TOKEN was one of the main reasons we built this feature. Will add `user_context.get_provider_tokens()` in this same PR and ship it today.

Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py]: > _This reply was posted by OpenHands AI on behalf of @xingyaoww, based on our Mar 17 sync meeting notes._

Thanks for pointing to the exact method! Will wire this up today.

Comment (xingyaoww) [reply openhands/app_server/sandbox/sandbox_router.py]: Implemented in fe0bbb3. The sandbox secrets endpoints now include provider tokens resolved lazily:

- `list_secret_names` includes provider token env var names (e.g. `github_token`) alongside custom secrets
- `get_secret_value` checks custom secrets first, then falls back to `get_provider_token_env_vars()` which resolves the latest token per-request (no snapshot)
- Added `AuthUserContext.get_provider_token_env_vars()` that maps `ProviderType` → env key and resolves fresh token values
- 6 new tests covering the provider token flow

File: openhands/app_server/user/user_router.py (hunk 11 above)
Comment (xingyaoww) [reply openhands/app_server/user/user_router.py:32]: This is still in `sandbox_router.py` under the sandbox prefix. The routes are already namespaced as `GET /sandboxes/{id}/settings/secrets` — they live in the sandbox router alongside other sandbox endpoints. Did you mean a separate file/module entirely, or is the current placement OK?

**Comment by xingyaoww** [comment]
## ✅ E2E Testing Complete (2026-03-17)

Deployed this PR to staging via [deploy#3436](https://github.com/OpenHands/deploy/pull/3436) (`ohpr-13383-240.staging.all-hands.dev`) at commit `c0ed13f` and ran the SDK e2e test from the companion [SDK PR #2409](https://github.com/OpenHands/software-agent-sdk/pull/2409).

### Results — all passing

| Phase | Status | Details |
|---|---|---|
| `workspace.get_llm()` | ✅ | LLM config inherited from SaaS (`litellm_proxy/minimax-m2.5` + api_key + base_url) |
| `workspace.get_secrets()` | ✅ | Returns **3 secrets**: `['secret_1', 'secret_2', 'github_token']` |
| Provider token discovery | ✅ | `github_token` exposed via sandbox secrets endpoint — **provider tokens now flow to SDK** |
| `LookupSecret` resolution | ✅ | All 3 secrets resolved: custom secrets (len=5 each) + `github_token` (len=40) |
| Sandbox cleanup | ✅ | Sandbox created + deleted cleanly |

### What was validated
- `GET /api/v1/users/me?expose_secrets=true` correctly returns LLM config
- `GET /api/v1/sandboxes/{id}/settings/secrets` returns both custom secrets AND provider tokens (`github_token`)
- `GET /api/v1/sandboxes/{id}/settings/secrets/{name}` resolves each secret value individually
- `get_provider_tokens(as_env_vars=True)` refactor works correctly — provider tokens appear as env-var-style names
- Server unit tests: 30/30 passing

Test artifacts in SDK PR: [`.pr/README.md`](https://github.com/OpenHands/software-agent-sdk/blob/feat/cloud-workspace-get-llm-secrets/.pr/README.md)

**Comment by xingyaoww** [comment]
## 📋 Full agent example stdout uploaded to SDK PR

Ran the SDK's `examples/02_remote_agent_server/10_cloud_workspace_saas_credentials.py` end-to-end against this PR's staging deployment (`ohpr-13383-240.staging.all-hands.dev`, commit `c0ed13f`).

Full stdout: [`example_10_full_stdout.log`](https://github.com/OpenHands/software-agent-sdk/blob/feat/cloud-workspace-get-llm-secrets/.pr/logs/example_10_full_stdout.log)

**Key results:**
- `get_llm()` → `litellm_proxy/minimax-m2.5` + api_key + base_url inherited
- `get_secrets()` → `['secret_1', 'secret_2', 'github_token']` (custom secrets + provider token)
- Agent resolved all 3 secrets as env vars inside sandbox, wrote `SECRETS_CHECK.txt`
- Conversation: 17 events, 34.2s, sandbox cleaned up on exit

**Comment by xingyaoww** [comment]
Confirmed that you can indeed get the secrets via API - gonna merge this

image

**Comment by enyst** [comment]
To note, from PR description
> expose_secrets=true is gated behind dual authentication:
> - Bearer token (proves identity via OPENHANDS_API_KEY)
> - X-Session-API-Key header (proves the caller has an active sandbox)
> The server validates that the session key maps to a sandbox owned by the Bearer token user.
> 
> A leaked API key alone cannot extract raw LLM credentials — an ephemeral, sandbox-scoped session key is also required.

Sorry to make this clear, but this is not accurate, as I was trying to point out in the related discussion.

If an `OPENHANDS_API_KEY` is leaked, then it gives access to the `X-Session-API-Key` too.

So it’s not gated behind dual auth. It’s behind only `OPENHANDS_API_KEY`. 🤔

Maybe in itself that’s not very bad, we just should be aware… The threat model here is: if `OPENHANDS_API_KEY` is leaked, then it gives access to the actual secrets, including LLM key, if I understand correctly? Sorry for the slow follow up here.

**Comment by xingyaoww** [comment]
https://github.com/OpenHands/OpenHands/pull/13383#issuecomment-4076659120

> If an OPENHANDS_API_KEY is leaked, then it gives access to the X-Session-API-Key too.

@enyst yes, but then the attacker would need to go into the cloud, spin a new sandbox, and get the LLM_API_KEY there. The action of spin a new sandbox should give us slightly more observability?

@OpenHands do you have any suggestion here? Maybe all we need is some audit layer to record all the access to the request of "secrets" / "llm settings" eventually? What's the better approach here

**Comment by xingyaoww** [comment]
@enyst's point is correct — the "dual auth" effectively collapses to single-factor since `OPENHANDS_API_KEY` can derive `X-Session-API-Key` by creating a sandbox. The current mitigation is that this derivation requires an observable action (sandbox creation), but that's weak on its own.

Here's a layered approach, ordered by effort/impact:

### 1. Structured audit logging on secrets access (immediate, high value)

Every call to `expose_secrets=true`, `GET .../secrets/{name}`, and `GET .../secrets` should emit a structured log event with:
- `user_id`, `sandbox_id`, `secret_name` (for individual fetches), `client_ip`, `timestamp`
- A dedicated log category (e.g., `openhands.security.secrets_access`) so it's easy to filter in Datadog/CloudWatch

This is the minimum viable audit trail and can be added with just a few `_logger.info(...)` calls in `user_router.py` and `sandbox_router.py`. The existing `_logger.warning` for ownership mismatch is a good start — we just need the happy-path logging too.

### 2. PostHog tracking for secrets access events (low effort)

PostHog is already integrated in the enterprise auth flow (`enterprise/server/routes/auth.py`). Adding a `posthog.capture()` for `secrets_accessed` events would enable:
- Anomaly detection (unusual access patterns per user)
- Dashboard visibility into how often SDK credential inheritance is actually used
- Alerting on suspicious spikes

### 3. Rate limiting on secrets endpoints (medium effort)

The `check_rate_limit_by_user_id` infrastructure already exists in `enterprise/server/utils/rate_limit_utils.py`. Applying it to secrets endpoints would cap how quickly an attacker with a leaked key can exfiltrate credentials — e.g., max 5 secrets lookups per minute per user.

### 4. Scope-gated session keys (longer term, strongest mitigation)

The strongest fix for the single-factor collapse would be to make secrets access an **opt-in capability** on the session key itself. For example:
- When the SDK creates a sandbox, it requests `scope=secrets` explicitly
- The session key is minted with that scope, and the secrets endpoints check for it
- This means a generic sandbox session key (e.g., from a webhook-triggered run) **cannot** access secrets even if the `OPENHANDS_API_KEY` is compromised

This is more invasive but directly addresses @enyst's concern — it decouples "I can run code in a sandbox" from "I can read the user's raw credentials."

### Recommendation

I'd suggest starting with **(1) + (2)** as an immediate follow-up — they're low-effort and give full visibility. **(3)** is a good next step. **(4)** can be designed in parallel as a longer-term hardening measure.

Want me to open a follow-up PR implementing the audit logging (items 1 & 2)?

**Comment by enyst** [comment]
@openhands Read https://github.com/OpenHands/OpenHands/pull/13383#issuecomment-4079680190 and understand all the comments in this PR and its behavior.

I think a single bearer auth is fine; we should think of follow-ups though. Go further than the above comment; I want you to make a wide investigation on whether we could also consider a sort of sub-identity for an AI agent running under an user account and how would that look like; what's the minimal version of that. WDYT?

Investigate the repo all you like. You have time.

Open an issue with your proposals.
